src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/DictionarySlim.cs

```diff
@@ -0,0 +1,380 @@
+
+        public int Count { get; private set; }
+
+        public int Capacity => _entries.Length;
+
+        public bool ContainsKey(TKey key)
+        {
+            Entry[] entries = _entries;
+            int entryIndex = GetEntryIndex((int)(GetHashCode(key) % (uint)entries.Length));
+
+            while (entryIndex != -1)
+            {
+                if (entries[entryIndex].key.Equals(key))
+                {
+                    return true;
+                }
+                entryIndex = entries[entryIndex].next;
+            }
+
+            return false;
+        }
+
+        public TValue GetValueOrDefault(TKey key)
+        {
+            Entry[] entries = _entries;
+            int entryIndex = GetEntryIndex((int)(GetHashCode(key) % (uint)entries.Length));
+
+            while (entryIndex != -1)
+            {
+                if (entries[entryIndex].key.Equals(key))
+                {
+                    return entries[entryIndex].value;
+                }
+                entryIndex = entries[entryIndex].next;
+            }
+
+            return default;
+        }
+
+        public bool Remove(TKey key)
+        {
+            Entry[] entries = _entries;
+            int bucketIndex = (int)(GetHashCode(key) % (uint)entries.Length);
+            int entryIndex = GetEntryIndex(bucketIndex);
+            
+            int lastIndex = -1;
+            while (entryIndex != -1)
+            {
+                if (entries[entryIndex].key.Equals(key))
+                {
+                    if (lastIndex != -1)
+                    {   // Fixup preceding element in chain to point to next (if any)
+                        entries[lastIndex].next = entries[entryIndex].next;
+                    }
+                    else
+                    {   // Fixup bucket to new head (if any)
+                        _buckets[bucketIndex] = entries[entryIndex].next + 1;
+                    }
+
+                    entries[entryIndex] = default; // could use RuntimeHelpers.IsReferenceOrContainsReferences
+
+                    entries[entryIndex].next = _freeList; // New head of free list
+                    _freeList = entryIndex;
+
+                    Count--;
+                    return true;
+                }
+                lastIndex = entryIndex;
+                entryIndex = entries[entryIndex].next;
+            }
+
+            return false;
+        }
+
+        public ref TValue this[TKey key]
+        {
+            get
+            {
+                Entry[] entries = _entries;
+                int bucketIndex = (int)(GetHashCode(key) % (uint)entries.Length);
+                int entryIndex = GetEntryIndex(bucketIndex);
+
+                for(int i = entryIndex; i != -1; i = entries[i].next)
+                {
+                    if (entries[i].key.Equals(key))
+                    {
+                        return ref entries[i].value;
+                    }
+                }
+
+                if (_freeList != -1)
```

Comment:
The insert code here is relatively large and generally going to be called much less often than retrieving existing items. It might help performance a bit to split it out into a separate function with `NoInlining` to reduce the number of registers spilled to the stack on the hot path (and maybe add `AggressiveInlining` on the indexer as well)

Reply:
Works well, added.